### PR TITLE
Replace os.system and remove functions with shell=True

### PR DESCRIPTION
**Pull request by maipbui**
Signed-off-by: maipbui <maibui@microsoft.com>

#### What I did
1. `subprocess()` - when using with `shell=True` is dangerous. Using subprocess function without a static string can lead to command injection.
2. `getstatusoutput` implementation has `shell=True`
3. `os` - not secure against maliciously constructed input and dangerous if used to evaluate dynamic content.
#### How I did it
1. `subprocess()` - use `shell=False` instead, use list of strings Ref: [https://semgrep.dev/docs/cheat-sheets/python-command-injection/#mitigation](https://semgrep.dev/docs/cheat-sheets/python-command-injection/#mitigation)
2. Use `getstatusoutput_noshell` function from sonic_py_common
3. `os` - use with `subprocess` function
#### How to verify it
Pass UT and manual test

### CODE REVIEW

File: config/main.py [hunk 1]
```diff
@@ -5628,7 +5628,7 @@ def dropcounters():
 
 
 #
-# 'install' subcommand ('config dropcounters install')
+# ' + argsinstall' subcommand ('config dropcounters install')
```
Comment (qiluo-msft) [inline config/main.py:5631]: >argsinstall [](http://example.com/codeflow?start=6&length=11)

Is this change intended?

File: debug/main.py [hunk 2]
```diff
@@ -35,89 +35,83 @@ def bgp():
...
     @bgp.command()
     @click.argument('prefix_or_iface', required=False)
     def keepalives(prefix_or_iface):
         """BGP Neighbor Keepalives"""
-        command = 'sudo vtysh -c "debug bgp keepalives'
+        command = ['sudo', 'vtysh', '-c', "debug bgp keepalives"]
         if prefix_or_iface is not None:
-            command += " " + prefix_or_iface
-        command += '"'
+            command = ['sudo', 'vtysh', '-c', "debug bgp keepalives " + prefix_or_iface]
```
Comment (qiluo-msft) [inline debug/main.py:63]: >['sudo', 'vtysh', '-c', "debug bgp keepalives " [](http://example.com/codeflow?start=22&length=47)

The sub array repeats in code. Better to define an variable.

File: debug/main.py [hunk 3]
```diff
@@ -35,89 +35,83 @@ def bgp():
...
     @bgp.command('neighbor-events')
     @click.argument('prefix_or_iface', required=False)
     def neighbor_events(prefix_or_iface):
         """BGP Neighbor Events"""
-        command = 'sudo vtysh -c "debug bgp neighbor-events'
+        command = ['sudo', 'vtysh', '-c', "debug bgp neighbor-events"]
         if prefix_or_iface is not None:
-            command += " " + prefix_or_iface
-        command += '"'
+            command = ['sudo', 'vtysh', '-c', "debug bgp neighbor-events " + prefix_or_iface]
```
Comment (qiluo-msft) [inline debug/main.py:72]: >'sudo', 'vtysh', '-c', "debug bgp neighbor-events  [](http://example.com/codeflow?start=23&length=50)

The same.

File: debug/main.py (hunk 2 above)
Comment (qiluo-msft) [reply debug/main.py]: Sorry for misleading. I am talking about specific issue that `'sudo', 'vtysh', '-c', "debug bgp keepalives "` is repeated twice. Not to mean a general issue.

However I see you take effort to improve the whole file, and it's better than what I was asking for.

Still this `vtysh_cmd + ["debug bgp keepalives "` repeated twice.

File: show/platform.py [hunk 4]
```diff
@@ -104,7 +104,8 @@ def psustatus(index, json, verbose):
 def ssdhealth(device, verbose, vendor):
     """Show SSD Health information"""
     if not device:
-        device = os.popen("lsblk -o NAME,TYPE -p | grep disk").readline().strip().split()[0]
+        _, stdout = getstatusoutput_noshell_pipe(["lsblk", "-o", "NAME,TYPE", "-p"], ["grep", "disk"])
+        device = stdout.strip().split()[0]
```
Comment (qiluo-msft) [inline show/platform.py:108]: >strip [](http://example.com/codeflow?start=24&length=5)

Do you change the behavior? Originally it only read first line, strip, and pick the first token.

File: sonic_package_manager/service_creator/creator.py [hunk 5]
```diff
@@ -97,7 +97,7 @@ def remove_if_exists(path):
     log.info(f'removed {path}')
 
 
-def run_command(command: str):
```
Comment (qiluo-msft) [inline sonic_package_manager/service_creator/creator.py:100]: >str [](http://example.com/codeflow?start=25&length=3)

Instead of removing the type hint, replace with a proper one?

File: undebug/main.py [hunk 6]
```diff
@@ -2,8 +2,9 @@
 import subprocess
 
 def run_command(command, pager=False):
-    click.echo(click.style("Command: ", fg='cyan') + click.style(command, fg='green'))
-    p = subprocess.Popen(command, shell=True, text=True, stdout=subprocess.PIPE)
+    command_str = ' '.join(command)
```
Comment (qiluo-msft) [inline undebug/main.py:5]: >join [](http://example.com/codeflow?start=22&length=4)

Is it better to use `shlex.join`? This will be applicable to multiple places.

File: show/platform.py [hunk 7]
```diff
@@ -104,7 +104,9 @@ def psustatus(index, json, verbose):
 def ssdhealth(device, verbose, vendor):
     """Show SSD Health information"""
     if not device:
-        device = os.popen("lsblk -o NAME,TYPE -p | grep disk").readline().strip().split()[0]
+        _, stdout = getstatusoutput_noshell_pipe(["lsblk", "-o", "NAME,TYPE", "-p"], ["grep", "disk"])
+        stdout_first_line = stdout.split('\n')[0]
```
Comment (qiluo-msft) [inline show/platform.py:108]: >[0] [](http://example.com/codeflow?start=46&length=3)

If stdout is empty, index 0 will be out of range, so there is risk of exception?

File: undebug/main.py (hunk 6 above)
Comment (maipbui) [reply undebug/main.py]: I have a concern, `shlex.join` is available starting Python 3.8, I don't think we can use this

**Comment by qiluo-msft** [comment]
>     clicommon.run_command(cmd + options, display_cmd=verbose)

Use string array instead of string for `run_command`.

---
Refers to: show/platform.py:116 in 620e25b. [](commit_id = 620e25bc27211a50fc456d908860799c68b72ed0, deletion_comment = False)

File: clear/main.py [hunk 8]
```diff
@@ -428,7 +426,7 @@ def clear_vlan_fdb(vlanid):
 @click.option('--devicename', '-d', is_flag=True, help="clear by name - if flag is set, interpret target as device name instead")
 def line(target, devicename):
     """Clear preexisting connection to line"""
-    cmd = "consutil clear {}".format("--devicename " if devicename else "") + str(target)
+    cmd = ["consutil", "clear", "{}".format("--devicename" if devicename else ""), str(target)]
```
Comment (qiluo-msft) [inline clear/main.py:429]: >"{}".format [](http://example.com/codeflow?start=32&length=11)

No need to format.

**Comment by qiluo-msft** [comment]
>     if value is None:

This is a valid bug fix. Since it is another topic and need other reviewers, could you separate it into another PR?

---
Refers to: config/main.py:1114 in 620e25b. [](commit_id = 620e25bc27211a50fc456d908860799c68b72ed0, deletion_comment = False)

**Comment by qiluo-msft** [comment]
>     if asic_type == 'cisco-8000':

could you separate it into another PR?

---
Refers to: config/main.py:1182 in 620e25b. [](commit_id = 620e25bc27211a50fc456d908860799c68b72ed0, deletion_comment = False)

**Comment by qiluo-msft** [comment]
>     clicommon.run_command(cmd + options, display_cmd=verbose)

The comment is still applicable.

---
In reply to: [1320111390](https://github.com/sonic-net/sonic-utilities/pull/2402#issuecomment-1320111390) [](http://example.com/codeflow?ancestors=1320111390,1320111390)

---
Refers to: show/platform.py:116 in 620e25b. [](commit_id = 620e25bc27211a50fc456d908860799c68b72ed0, deletion_comment = False)

**Comment by maipbui** [comment]
> > ```
> > clicommon.run_command(cmd + options, display_cmd=verbose)
> > ```
> 
> The comment is still applicable.
> 
> In reply to: [1320111390](https://github.com/sonic-net/sonic-utilities/pull/2402#issuecomment-1320111390)
> 
> Refers to: show/platform.py:116 in [620e25b](https://github.com/sonic-net/sonic-utilities/commit/620e25bc27211a50fc456d908860799c68b72ed0). [](commit_id = [620e25b](https://github.com/sonic-net/sonic-utilities/commit/620e25bc27211a50fc456d908860799c68b72ed0), deletion_comment = False)

I'm working on it

File: consutil/lib.py [hunk 9]
```diff
@@ -327,14 +330,23 @@ def _parse_processes_info(output):
...
         output = proc.stdout.read()
         error = proc.stderr.read()
         if abort and error != "":
             click.echo("Command resulted in error: {}".format(error))
             sys.exit(ERR_CMD)
         return output if abort else (output, error)
 
+    @staticmethod
+    def run_command_pipe(*args, abort=True):
+        """runs command, exit if stderr is written to and abort argument is ture, returns stdout, stderr otherwise"""
```
Comment (qiluo-msft) [inline consutil/lib.py:343]: >ture [](http://example.com/codeflow?start=76&length=4)

typo

File: consutil/lib.py [hunk 10]
```diff
@@ -327,14 +330,23 @@ def _parse_processes_info(output):
...
+        exitcodes, output = getstatusoutput_noshell_pipe(*args)
+        if abort and any(exitcodes) and output != '':
```
Comment (qiluo-msft) [inline consutil/lib.py:345]: >output [](http://example.com/codeflow?start=40&length=6)

The original function `run_command` is checking stderr, but you are checking stdout. Did you change the behavior?

**Comment by qiluo-msft** [comment]
>     def build_header(self, pg_drop_type):

What is the motivation to change this function? Is it related to the PR title?

---
Refers to: scripts/pg-drop:109 in aab1aa9. [](commit_id = aab1aa96a7a1012e9e878e34ddd637ecb53b62b3, deletion_comment = False)

File: consutil/lib.py (hunk 10 above)
Comment (maipbui) [reply consutil/lib.py:345]: in case `getstatusoutput_noshell_pipe` has errors or not, it both writes to stdout

**Comment by maipbui** [comment]
> > ```
> > def build_header(self, pg_drop_type):
> > ```
> 
> What is the motivation to change this function? Is it related to the PR title?
> 
> Refers to: scripts/pg-drop:109 in [aab1aa9](https://github.com/sonic-net/sonic-utilities/commit/aab1aa96a7a1012e9e878e34ddd637ecb53b62b3). [](commit_id = [aab1aa9](https://github.com/sonic-net/sonic-utilities/commit/aab1aa96a7a1012e9e878e34ddd637ecb53b62b3), deletion_comment = False)

I didn't change this function. In this file, I changed the use of pickle module to json module per Semgrep recommendation.

File: config/main.py [hunk 11]
```diff
@@ -633,10 +634,12 @@ def _remove_bgp_neighbor_config(config_db, neighbor_ip_or_hostname):
 def _change_hostname(hostname):
     current_hostname = os.uname()[1]
     if current_hostname != hostname:
-        clicommon.run_command('echo {} > /etc/hostname'.format(hostname), display_cmd=True)
-        clicommon.run_command('hostname -F /etc/hostname', display_cmd=True)
-        clicommon.run_command(r'sed -i "/\s{}$/d" /etc/hosts'.format(current_hostname), display_cmd=True)
-        clicommon.run_command('echo "127.0.0.1 {}" >> /etc/hosts'.format(hostname), display_cmd=True)
+        with open('etc/hostname', 'w') as f:
```
Comment (qiluo-msft) [inline config/main.py:637]: >etc [](http://example.com/codeflow?start=19&length=3)

It should be absolute path. Did you test?

File: config/main.py [hunk 12]
```diff
@@ -633,10 +634,12 @@ def _remove_bgp_neighbor_config(config_db, neighbor_ip_or_hostname):
...
+            f.write(str(hostname) + '\n')
+        clicommon.run_command(['hostname', '-F', '/etc/hostname'], display_cmd=True)
+        clicommon.run_command(['sed', '-i', "/\s{}$/d".format(current_hostname), '/etc/hosts'], display_cmd=True)
+        with open('etc/hosts', 'a') as f:
```
Comment (qiluo-msft) [inline config/main.py:641]: >etc [](http://example.com/codeflow?start=19&length=3)

The same

File: config/main.py (hunk 11 above)
Comment (maipbui) [reply config/main.py:637]: My mistake, will fix.
